service/graph.go

```diff
@@ -31,21 +32,236 @@ var _ pipelines = (*pipelinesGraph)(nil)
 type pipelinesGraph struct {
 	// All component instances represented as nodes, with directed edges indicating data flow.
 	componentGraph *simple.DirectedGraph
+
+	// Keep track of how nodes relate to pipelines, so we can declare edges in the graph.
+	pipelines map[component.ID]*pipelineNodes
+}
+
+func buildPipelinesGraph(ctx context.Context, set pipelinesSettings) (pipelines, error) {
+	pipelines := &pipelinesGraph{
+		componentGraph: simple.NewDirectedGraph(),
+		pipelines:      make(map[component.ID]*pipelineNodes, len(set.PipelineConfigs)),
+	}
+	for pipelineID := range set.PipelineConfigs {
+		pipelines.pipelines[pipelineID] = &pipelineNodes{
+			receivers: make(map[int64]graph.Node),
+			exporters: make(map[int64]graph.Node),
+		}
+	}
+
+	if err := pipelines.createNodes(set); err != nil {
+		return nil, err
+	}
+	pipelines.createEdges()
+	if err := pipelines.buildNodes(ctx, set); err != nil {
+		return nil, err
+	}
+	return pipelines, nil
+}
+
+// Creates a node for each instance of a component and adds it to the graph
+func (g *pipelinesGraph) createNodes(set pipelinesSettings) error {
+
+	// map[connectorID]pipelineIDs
+	// Keep track of connectors and where they are used.
+	connectorsAsExporter := make(map[component.ID][]component.ID)
+	connectorsAsReceiver := make(map[component.ID][]component.ID)
+
+	for pipelineID, pipelineCfg := range set.PipelineConfigs {
+		for _, recvID := range pipelineCfg.Receivers {
+			if set.Connectors.IsConfigured(recvID) {
+				connectorsAsReceiver[recvID] = append(connectorsAsReceiver[recvID], pipelineID)
+				continue
+			}
+			g.addReceiver(pipelineID, recvID)
+		}
+		for _, procID := range pipelineCfg.Processors {
+			g.addProcessor(pipelineID, procID)
+		}
+		for _, exprID := range pipelineCfg.Exporters {
+			if set.Connectors.IsConfigured(exprID) {
+				connectorsAsExporter[exprID] = append(connectorsAsExporter[exprID], pipelineID)
+				continue
+			}
+			g.addExporter(pipelineID, exprID)
+		}
+	}
+
+	if len(connectorsAsExporter) != len(connectorsAsReceiver) {
+		return fmt.Errorf("each connector must be used as both receiver and exporter")
+	}
```

Comment:
Is this not checked in the "Validation" of the config? Why duplicate the logic here?

Reply:
You're right. This was just defensive code, but it is unnecessary. I will remove.